syntax = "proto3";

service SmartObjectUpdate {
  rpc UpdateObject (SmartObjectDetails) returns (UpdateObjectReply) {}
}

message UpdateObjectReply {
  bool success = 1;
}

message SmartObjectsList {
  repeated SmartObjectDetails objects = 1;
}

message SmartObjectDetails {
  bool status = 1;
  string ip = 2;
  int32 port = 3;
  string id = 8;

  // The details for the smart object can be one of these (only one); These are the currently supported objects
  oneof object_details {
    LightDetails light = 4;
    LuminositySensor luminosity_sensor = 5;
    AirConditionerDetails ac = 6;
    TemperatureSensorDetails temp_sensor = 7;
    SmokeSensor smoke_sensor = 10;
    Sprinkler sprinkler = 9;
  }
}

message LightDetails {
  string name = 1;
  string color = 2;
}

message LuminositySensor {
  string name = 1;
  float reading = 2;
}

message SmokeSensor {
  string name = 1;
  float reading = 2;
}

message Sprinkler {
  string name = 1;
}

enum AirConditionerMode {
  FAN_MODE = 0;
  SILENT_MODE = 1;
  EXTREME_CHILL = 2;
}

message AirConditionerDetails {
  int32 temperature = 1;
  AirConditionerMode mode = 2;
}

message TemperatureSensorDetails {
  int32 temperature = 1;
}